### DISCUSSION TITLE
Need help with InternalError: unimplemented in VstSsaGen.visitFuncExpr()

### DISCUSSION THREAD

**Discussion by gmlewis** [titzer/virgil#255] [Q&A] [answered]
I'm trying to learn Virgil by writing an Extism PDK for it here:
https://github.com/gmlewis/virgil/tree/extism-pdk/apps/Extism

I've got the "greet" Virgil PDK example working, but am running into a problem with the `count-vowels` example:

```
$ ./build.sh
...
+ v3c -entry-export=_initialize -heap-size=500m -main-export=_initialize -output=count-vowels -target=wasm count-vowels/CountVowels.v3 ../../lib/pdk/Config.v3 ../../lib/pdk/Extism.v3 ../../lib/pdk/Host.v3 ../../lib/pdk/Http.v3 ../../lib/pdk/Memory.v3 ../../lib/pdk/Var.v3 ../../lib/util/Arrays.v3 ../../lib/util/BitMatrix.v3 ../../lib/util/Chars.v3 ../../lib/util/DataReader.v3 ../../lib/util/DataWriter.v3 ../../lib/util/Edge.v3 ../../lib/util/Functions.v3 ../../lib/util/GlobMatcher.v3 ../../lib/util/IO.v3 ../../lib/util/Ints.v3 ../../lib/util/Layouts.v3 ../../lib/util/List.v3 ../../lib/util/Longs.v3 ../../lib/util/Map.v3 ../../lib/util/Option.v3 ../../lib/util/Poly.v3 ../../lib/util/Random.v3 ../../lib/util/Ranges.v3 ../../lib/util/Stack.v3 ../../lib/util/StringBuilder.v3 ../../lib/util/Strings.v3 ../../lib/util/TextReader.v3 ../../lib/util/Token.v3 ../../lib/util/UID.v3 ../../lib/util/Utf8.v3 ../../lib/util/Vector.v3 ../../lib/util/WorkQueue.v3
InternalError: in CountVowels.count_vowels: unimplemented in VstSsaGen.visitFuncExpr()
```

I poked around in https://github.com/gmlewis/virgil/blob/89aefcdcae2eb044b47a5fc2a022075c07fccd5e/aeneas/src/ssa/VstSsaGen.v3#L1179

but am wondering why Ssa is involved at all when generating a `.wasm` file...
and I'm completely at a loss as to how to run a debugger on the compiler.

Any help would be greatly appreciated.  Thanks.

**Comment by titzer** [comment, marked as answer]
Ah ha, looks like you have a nested function here: https://github.com/gmlewis/virgil/blob/extism-pdk/apps/Extism/count-vowels/CountVowels.v3#L51

That feature is a work-in-progress. I should probably put it behind a flag.

**Comment by titzer** [comment]
> am wondering why Ssa is involved at all when generating a `.wasm` file..

All code goes from source -> VST -> SSA, and then backends for each target process SSA a little differently. E.g. the Wasm backend goes through machine lowering (SSA -> SSA transformation) and then translates SSA into Wasm. The above error occurred at the VST -> SSA translation stage (thus the `VstSsaGen.v3` filename).

**Comment by gmlewis** [comment]
Awesome, thank you, @titzer !  I'll move it.  😄
